### PR TITLE
Fallback to xcode-select with xcode_locator

### PR DESCRIPTION
**Pull request by keith**
On machines that have spotlight disabled for build performance reasons,
LSCopyApplicationURLsForBundleIdentifier can return nothing, even when
Xcode is installed. Previously This would result in xcode_locator
failing, and the build eventually failing if you relied on Xcode. Now we
fall back to the xcode-selected Xcode version and use that.

### CODE REVIEW

**Comment by dflems** [comment]
I think `xcode-select -p` will return the path to CommandLineTools (if Xcode is not installed) (`/Library/Developer/CommandLineTools`). In that event, your method will return `/Library`.

It seems like `xcode-select -p` does a few things in order:

* Check `DEVELOPER_DIR` environment variable
* Check `/var/db/xcode_select_link` symlink
* Check `/usr/share/xcode-select/xcode_dir_link` symlink
* Check for contents of file at `/usr/share/xcode-select/xcode_dir_path`
* Checks explicitly for `/Applications/Xcode.app/Contents/Developer`
* Checks for command line tools known location (`/Library/Developer/CommandLineTools`)

**Comment by keith** [comment]
> I think xcode-select -p will return the path to CommandLineTools (if Xcode is not installed)

This will end up being ok because this script does some post processing to validate that the array of directories are all valid Xcode's by checking their bundle IDs. So if it returns the command line tools that should get filtered out.

**Comment by kastiglione** [comment]
> It seems like `xcode-select -p` does a few things in order

@dflems  as an aside, how did you get this list, especially the _in order_ part.

**Comment by keith** [comment]
It actually looks like this isn't enough, the xcode-locator path is hard coded in another place https://github.com/keith/bazel/blob/ecee40d949d909a3eca9c5514239efa86cf1a64b/tools/cpp/osx_cc_configure.bzl#L69

**Comment by dflems** [comment]
@kastiglione I opened `/usr/lib/libxcselect.dylib` in Hopper. The function `xcselect_get_developer_dir_path` contains this logic.

**Comment by keith** [comment]
Actually if this was merged, this would solve both cases, we just can't use this locally without forking bazel.

**Comment by keith** [comment]
Our problem here is that on our third party CI provider we have VMs with spotlight disabled, so this command currently fails. We also have those machines setup with only a single version, but even if they had multiple, I think it's safe to assume we'd set the correct version to xcode-select. Although this could still always end up conflicting with what you pass for `--xcode-version`, but at least there's a possibility that you could find the right version

**Comment by kastiglione** [comment]
> Bazel will only be able to find the active `xcode-select`'ed version. With this in mind, we should probably print a warning when spotlight is disabled

When more than one Xcode is found, the one that's chosen is the one `xcode-select` points to. Given this, I don't think a warning is of use.

**Comment by DavidGoldman** [comment]
> > Bazel will only be able to find the active `xcode-select`'ed version. With this in mind, we should probably print a warning when spotlight is disabled
> 
> When more than one Xcode is found, the one that's chosen is the one `xcode-select` points to. Given this, I don't think a warning is of use.

You can imagine a case where both Xcode 9 and Xcode 10 are installed. Xcode 10 is the active Xcode version, but the user passes `--xcode_version=9` on the command line. If spotlight is disabled, it won't work until the user either drops the flag or swaps the active Xcode version and does a bazel clean.

**Comment by keith** [comment]
>  If spotlight is disabled, it won't work until the user either drops the flag or swaps the active Xcode version and does a bazel clean.

This is definitely the case. The good news is this is the state today as well, regardless of this patch, except the error you get ends up being much more opaque, this is what we got:

```
java.lang.IllegalStateException: //deps/generated/pb/api/models/v1/example_model_2:DTOExampleModel2V1Framework BuildConfigurationValue.Key[7efd4f2fe9d89ca30f400c7605ce9bff] false -> ErrorInfo{exception=com.google.devtools.build.lib.analysis.config.InvalidConfigurationException: No toolchain found for --cpu='ios_i386' --compiler='compiler'. Valid toolchains are: [
  stub_armeabi-v7a: --cpu='armeabi-v7a' --compiler='compiler',
  local: --cpu='darwin' --compiler='compiler',
  msys_x64_mingw: --cpu='x64_windows' --compiler='mingw-gcc',
  msvc_x64: --cpu='x64_windows' --compiler='msvc-cl',
], rootCauses={ConfigurationFragmentKey(class=com.google.devtools.build.lib.rules.cpp.CppConfiguration, checksum=f52987aec1db55fc58eb0c4f8e275a42)}, cycles=[], isCatastrophic=false, rootCauseOfException=ConfigurationFragmentKey(class=com.google.devtools.build.lib.rules.cpp.CppConfiguration, checksum=f52987aec1db55fc58eb0c4f8e275a42), isDirectlyTransient=false, isTransitivelyTransient=false}
	at com.google.common.base.Preconditions.checkState(Preconditions.java:823)
	at com.google.devtools.build.lib.skyframe.SkyframeBuildView.assertSaneAnalysisError(SkyframeBuildView.java:539)
	at com.google.devtools.build.lib.skyframe.SkyframeBuildView.configureTargets(SkyframeBuildView.java:371)
	at com.google.devtools.build.lib.analysis.BuildView.update(BuildView.java:357)
	at com.google.devtools.build.lib.buildtool.AnalysisPhaseRunner.runAnalysisPhase(AnalysisPhaseRunner.java:212)
	at com.google.devtools.build.lib.buildtool.AnalysisPhaseRunner.execute(AnalysisPhaseRunner.java:121)
	at com.google.devtools.build.lib.buildtool.BuildTool.buildTargets(BuildTool.java:143)
	at com.google.devtools.build.lib.buildtool.BuildTool.processRequest(BuildTool.java:262)
	at com.google.devtools.build.lib.runtime.commands.BuildCommand.exec(BuildCommand.java:83)
	at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.execExclusively(BlazeCommandDispatcher.java:484)
	at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.exec(BlazeCommandDispatcher.java:204)
	at com.google.devtools.build.lib.server.GrpcServerImpl.executeCommand(GrpcServerImpl.java:870)
	at com.google.devtools.build.lib.server.GrpcServerImpl.access$2100(GrpcServerImpl.java:111)
	at com.google.devtools.build.lib.server.GrpcServerImpl$2.lambda$run$0(GrpcServerImpl.java:939)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```

**Comment by keith** [comment]
I'm happy to update this patch to add searching standard directories for things matching `Xcode.*\.app` if you think we should do that before this can land

**Comment by DavidGoldman** [comment]
Thanks for the extra context here; are you still interested in submitting this?

**Comment by keith** [comment]
I think this is still a reasonable fallback if this fails. But if you don't want to add the complexity that's fine with me.


--
Keith Smiley

File: tools/osx/xcode_locator.m [hunk 1]
```diff
@@ -122,12 +140,21 @@ static void AddEntryToDictionary(
...
   CFErrorRef cfError;
-  NSArray *array = CFBridgingRelease(LSCopyApplicationURLsForBundleIdentifier(
+  NSArray<NSURL *> *array =
+      CFBridgingRelease(LSCopyApplicationURLsForBundleIdentifier(
       cfBundleID, &cfError));
   if (array == nil) {
-    NSError *nsError = (__bridge NSError *)cfError;
-    fprintf(stderr, "error: %s\n", nsError.description.UTF8String);
-    return nil;
+    // If spotlight is disabled on the host machine for performance reasons
```
[the remark this answers is not in the document]
Comment (keith) [reply tools/osx/xcode_locator.m]: Updated

**Comment by jmmv** [comment]
Just saw this when reviewing the merge request.

I think we must handle the case where Spotlight is disabled and falling back to whatever `xcode-select` says seems reasonable.

I have mild concerns about the confusion that can arise from `--xcode_version` and whatever `xcode-select` does though.

Could we add some extra log messages to `xcode_locator` and then ensure that Bazel captures and logs those messages unconditionally to its own log? (I think it currently captures the messages but only displays them on failure.)

Additionally, could we can annotate the output of `xcode_locator` to mention where each path came from (Spotlight or `xcode-select`), change Bazel to parse that extra field, and use this information to provide better diagnostic messages when finding the version fails?

**Comment by keith** [comment]
I'm going to drop this change, I still think it's a good fallback but since we found our core issue where this was a problem I'm not super interested in finishing this up. If someone else would like to pick it up feel free!

**Pull request closed without merging**